New DateInput component

connor-baer opened the pull request:
Addresses [DSYS-656](https://sumupteam.atlassian.net/browse/DSYS-656).

## Purpose

[`input[type="date"]`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date) is a rare example of a native element that doesn't meet accessibility requirements and warrants a custom component. The native component also lacks customizability to disable specific dates or allow the selection of a date range.

## Approach and changes

- Replace the DateInput component previously based on the native date input with a custom implementation. Users can enter a date in two ways:
	- Type into the text box. The year, month, and day segments are split into individual inputs to improve usability, internationalization and validation.
	- Select on a calendar. Pressing the trailing calendar button opens the Calendar component (#2494) in a dialog.

## Definition of done

* [ ] Development completed
* [x] Reviewers assigned
* [ ] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

sirineJ commented on packages/circuit-ui/components/DateInput/DateInput.spec.tsx:
```diff
@@ -13,21 +13,288 @@
+      expect(screen.getByLabelText(/day/i)).toHaveAttribute('max', '15');
+    });
+
+    it('should allow users to select a date on a calendar', async () => {
```
TODO: add a test for the modal version

connor-baer commented on packages/circuit-ui/components/DateInput/DateInput.tsx:
```diff
@@ -15,79 +15,522 @@
+            />
+          </FieldLegend>
+          {/* biome-ignore lint/a11y/useKeyWithClickEvents: */}
+          <div className={classes.wrapper} onClick={handleClick}>
```
Yes, it enables users to press anywhere on the field to focus it. The input segments themselves have fairly small hit areas.

connor-baer commented on packages/circuit-ui/components/DateInput/components/Segment.tsx:
```diff
@@ -0,0 +1,70 @@
+        {...props}
+      />
+      <span ref={sizeRef} className={classes.size} aria-hidden="true">
+        {props.value || props.placeholder}
```
Good point about the layout shift 🤔

The issue with fixed-width segment inputs is that the spacing between them looks off when they only contain one digit. We could solve this by giving the container a minimum width, so the segments inputs inside can resize without affecting the surrounding elements.

connor-baer commented on packages/circuit-ui/vendor/dialog-polyfill/index.js:
```diff
@@ -621,7 +621,7 @@ if (typeof window === 'undefined') {
    */
   dialogPolyfill.forceRegisterDialog = function (element) {
-    if (window.HTMLDialogElement || element.showModal) {
+    if (window.HTMLDialogElement && element.showModal) {
```
`jsdom`, our unit test environment, implements `window.HTMLDialogElement` but not `element.showModal`. I wanted to silence the warning, but didn't mean to commit the change. I'll revert it.

connor-baer commented on packages/circuit-ui/components/DateInput/hooks.ts:
```diff
@@ -0,0 +1,334 @@
+
+        if (isNumber(year)) {
+          year = clamp(year, MIN_YEAR, MAX_YEAR);
+        }
```
We still need to update the internal state with the new field value(s), which require these calculations.

connor-baer commented on packages/circuit-ui/components/DateInput/hooks.ts:
```diff
@@ -0,0 +1,334 @@
+  const min = minDate ? minDate.year : 1;
+  const max = maxDate ? maxDate.year : 9999;
+
+  const onChange = (event: FormEvent<HTMLInputElement>) => {
```
AFAIR, I got the type from the native input element types. You're right though, `ChangeEvent` is more specific 👍🏻

connor-baer commented:
I've refactored the internals significantly:

- I merged the `use(Year|Month|Day)Segment` hooks into the `usePlainDateState` hook which is responsible for all date-related logic now. 
- I moved all keyboard interaction logic to the DateSegment component, which is a text input with `role="spinbutton"` now to support more keyboard shortcuts.

sirineJ commented on packages/circuit-ui/util/helpers.ts:
```diff
@@ -96,3 +99,17 @@ export function last<T>(array: T[]): T;
+  min: number, // inclusive
+  max: number, // inclusive
+) {
+  const modulus = max - min + 1;
```
should we throw is value is not in range ?

sirineJ commented on packages/circuit-ui/components/DateInput/components/DateSegment.tsx:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    // Don't allow editing the value when the input is disabled or read-only
+    if (input.disabled || input.readOnly) {
```
why do we need to be explicit about this if the input is disabled ?

connor-baer replied:
For some reason, the `onKeyDown` event handler is still called when the `input` disabled. Surprised me as well 😳

sirineJ commented on packages/circuit-ui/components/DateInput/components/DateSegment.tsx:
```diff
@@ -0,0 +1,197 @@
+    let newValue: number;
+
+    const getValue = (offset: number) =>
+      value ? shiftInRange(value, offset, min, max) : defaultValue;
```
👏

sirineJ commented on packages/circuit-ui/components/DateInput/hooks/usePlainDateState.ts:
```diff
@@ -0,0 +1,212 @@
+    month: {
+      value: values.month,
+      'aria-valuetext': values.month
+        ? [values.month, getMonthName(values.month, locale)].join(', ')
```
❤️

sirineJ commented on packages/circuit-ui/components/SidePanel/SidePanelContext.spec.tsx:
```diff
@@ -45,7 +45,7 @@ import {
 } from './SidePanelContext.js';
 
-vi.mock('../../hooks/useMedia');
+vi.mock('../../hooks/useMedia/index.js');
```
why so explicit ?

connor-baer replied:
The final ES module spec requires full file paths ([ref](https://nodejs.org/api/esm.html#mandatory-file-extensions)). I'm surprised that this worked before; I assume Vite does some magic under the hood to support ambiguous file paths.
